gb/src/joypad_terminal_menu.rs

```diff
@@ -0,0 +1,111 @@
+use std::{io::Write, time::Duration};
+
+use crossterm::{QueueableCommand, style::{self, Stylize}, cursor, terminal::{self, ClearType}, event::{self, KeyCode}};
+use lib_gb::keypad::{joypad_provider::JoypadProvider, joypad::Joypad, self, button::Button};
+
+pub struct TerminalRawModeJoypadProvider<F:Fn(KeyCode)->Option<Button>>{
+    mapper:F
+}
+
+impl<F:Fn(KeyCode)->Option<Button>> TerminalRawModeJoypadProvider<F> {
+    pub fn new(mapper:F)->Self{
+        Self{mapper}
+    }
+} 
+
+impl<F:Fn(KeyCode)->Option<Button>> JoypadProvider for TerminalRawModeJoypadProvider<F>{
+    fn provide(&mut self, joypad:&mut Joypad) {
+        terminal::enable_raw_mode().unwrap();
+        
+        joypad.buttons.fill(false);
+        // polling so this wont block
+        // even tough I can implement the keyboard for this to block I cant implement the rest of the joypads to block
+        // so the behavior of this module will have to assume that the input is non blocking
+        if event::poll(Duration::from_millis(100)).unwrap(){
+            if let event::Event::Key(event) = event::read().unwrap(){
+                let mapper = &(self.mapper);
+                if let Option::Some(button) = mapper(event.code){
+                    joypad.buttons[button as usize] = true;
+                }
+            }
+        }
+
+        terminal::disable_raw_mode().unwrap();
+    }
+}
+
+impl<F:Fn(KeyCode)->Option<Button>> Drop for TerminalRawModeJoypadProvider<F>{
+    fn drop(&mut self) {
+        // in case of panic in the middle of the raw mode 
+        terminal::disable_raw_mode().unwrap();
+    }
+}
+
+pub struct MenuOption<T>{
+    pub prompt:String,
+    pub value:T
+}
+
+pub struct JoypadTerminalMenu<T>{
+    options: Vec<MenuOption<T>>,
+    selection: usize
+}
+
+impl<T> JoypadTerminalMenu<T>{
+    pub fn new(options:Vec<MenuOption<T>>)->Self{
+        JoypadTerminalMenu{options,selection:0}
+    }
+
+    pub fn get_menu_selection<JP:JoypadProvider>(&mut self, provider:&mut JP)->&T{
+        let mut joypad = Joypad::default();
+        let mut redraw = true;
+        while !joypad.buttons[keypad::button::Button::A as usize]{
+            if redraw{
+                self.display_options();
+                redraw = false;
+            }
+            let prev_joypad = Joypad{buttons:joypad.buttons.clone()};
+            provider.provide(&mut joypad);
+            if joypad.buttons[keypad::button::Button::Up as usize] && !prev_joypad.buttons[keypad::button::Button::Up as usize]{
+                if self.selection > 0{
+                    self.selection -= 1;
+                    redraw = true;
+                }
+            }
+            else if joypad.buttons[keypad::button::Button::Down as usize] && !prev_joypad.buttons[keypad::button::Button::Down as usize]{
```

Comment:
Add comment explaining those ifs.

The problem with non blocking testing for the io

Reply:
Extracted to a function